app/scripts/lib/transaction/swaps.ts

```diff
@@ -0,0 +1,122 @@
+import { pickBy, merge } from 'lodash';
+import {
+  TransactionStatus,
+  TransactionType,
+  TransactionMeta,
+} from '../../../../shared/constants/transaction';
+import MetamaskController from '../../metamask-controller';
+
+type SwapOptions = {
+  // Whether this transaction required an approval transaction
+  hasApproveTx: boolean;
+  // Additional metadata to store for the transaction
+  meta: any;
+};
+
+type SwapApprovalTransaction = {
+  sourceTokenSymbol: string;
+  destinationTokenSymbol: string;
+  type: TransactionType;
+  destinationTokenDecimals: string;
+  destinationTokenAddress: string;
+  swapMetaData: string;
+  swapTokenValue: string;
+  estimatedBaseFee: string;
+  approvalTxId: string;
+};
+
+/**
+ * Creates a new transaction with swap-specific metadata
+ *
+ * @param this - MetamaskController
+ * @param swapOptions - Swap-specific options
+ * @param transactionType - The type of transaction
+ * @param transactionMeta - The transaction metadata
+ * @returns The updated transaction metadata
+ */
+export async function createSwapsTransaction(
+  this: MetamaskController,
+  swapOptions: SwapOptions,
+  transactionType: TransactionType,
+  transactionMeta: TransactionMeta,
+) {
+  if (
+    transactionType === TransactionType.swap &&
+    swapOptions?.hasApproveTx === false &&
+    transactionMeta.simulationFails
+  ) {
+    await this.txController.cancelTransaction(transactionMeta.id);
+    throw new Error('Simulation failed');
+  }
+
+  const swapsMeta = swapOptions?.meta;
+
+  if (!swapsMeta) {
+    return transactionMeta;
+  }
+
+  if (transactionType === TransactionType.swapApproval) {
+    this.swapsController.setApproveTxId(transactionMeta.id);
```

Comment:
Should we move the logic in these two `transactionType` conditions to separate private methods like `updateSwapApprovalTransaction`?

Reply:
Since we leave this functionality as it is, I will consider your comment while implementing on core.